pallets/nomad/da-bridge/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// Edit this file to define custom logic or remove it if it is not needed.
+/// Learn more about FRAME and the core library of Substrate FRAME pallets:
+/// <https://docs.substrate.io/v3/runtime/frame>
+pub use pallet::*;
+
+mod message;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use da_primitives::traits::ExtendedHeader;
+	use frame_support::{
+		pallet_prelude::*,
+		sp_runtime::{
+			traits::{CheckedSub, Header},
+			ArithmeticError,
+		},
+	};
+	use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+	use home::Pallet as Home;
+	use nomad_core::TypedMessage;
+	use primitive_types::H256;
+
+	use crate::message::{DABridgeMessages, DataRootMessage};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + home::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		#[pallet::constant]
+		type DABridgePalletId: Get<H256>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	// Block number to block hash mapping
+	#[pallet::storage]
+	#[pallet::getter(fn finalized_block_number_to_block_hash)]
+	pub type FinalizedBlockNumberToBlockHash<T: Config> =
+		StorageMap<_, Twox64Concat, T::BlockNumber, T::Hash>;
+
+	// Genesis config
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {}
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self { Self {} }
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {}
+	}
+
+	// Hooks
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(block_number: T::BlockNumber) {
```

Comment:
Is the block number given in the on_finalize hook _minus 1_ not guaranteed to be finalized?

Reply:
No, it's not. Only GRANDPA can guarantee the finalization. I can see 2 possible solutions here:
- Use an offchain worker, who is going to follow the chain finalization, or
- Integrate your pallet with the finality-grandpa gadget (I need to double-check if this is feasible in our time restriction)

Reply:
Final consensus from call is that we allow unfinalized block headers to be dispatched into home's merkle tree BUT we rely on fact that updater will only attest to final data. If the updater doesn't, it will be slashed for submitting an improper update (update that doesn't match actual on-chain state)